### PR TITLE
feat: update react-native-spokestack to 5@beta

### PR DESCRIPTION
**Pull request by timmywil**
- We were able to remove `rn-fetch-blob` completely, as well as remove all the model downloading code from the tray. RNS handles all that now. `@react-native-community/netinfo` was also removed.
- the nluModelUrls and wakewordModelUrls props have been changed
to nlu and wakeword respectively. They now accept source objects (e.g.
require('./filter.tflite')) as well as remote URLs
- addListener and removeListener have been changed to
addEventListener and removeEventListener, which are now direct exports
from react-native-spokestack. removeAllListeners has also been re-exported.
- SpokestackListenerEvent enum has been removed. Strings are used for events instead.
- Documentation has been updated to reflect these changes, bringing the tray docs in line with the latest react-native-spokestack.
- Sample app has been updated and tested on iOS and Android

TODO:

- [x] Fix reloading the tray in development mode.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -131,45 +141,86 @@ Set the AudioSession category, to enable microphone input and play from the spea
...
-1. You can disable Flipper imports without removing any code from the AppDelegate. To do this, open your xcworkspace file in XCode. Go to your target, then Build Settings, search for "C Flags", remove `-DFB_SONARKIT_ENABLED=1` from flags.
-1. Remove all Flipper-related code from your AppDelegate.m.
+<details>
+  <summary>Android details</summary>
 
-In our example app, we've done option 1 and left in the Flipper code in case they get it working in the future and we can add it back.
+### ASR Support
 
-## Android installation
+The example usage uses the system-provided ASRs (`AndroidSpeechRecognizer` and `AppleSpeechRecognizer`). However, `AndroidSpeechRecognizer` is not available on 100% of devices. If such a device is supported, use Spokestack ASR instead by setting the `profile` to a Spokestack profile using the `profile` prop.
```
Comment (space-pope) [inline README.md]:
```suggestion
The example usage uses the system-provided ASRs (`AndroidSpeechRecognizer` and `AppleSpeechRecognizer`). However, `AndroidSpeechRecognizer` is not available on 100% of devices. If your app supports a device that doesn't have built-in speech recognition, use Spokestack ASR instead by setting the `profile` to a Spokestack profile using the `profile` prop.
```

File: README.md [hunk 2]
```diff
@@ -204,14 +255,14 @@ export default function ConversationHandler({ navigation }) {
...
-      // The NLU models are downloaded once
+      // The NLU models are downloaded and then cached
       // when the app is first installed.
-      // See https://spokestack.io/docs/Concepts/nlu
-      // for descriptions of these files.
-      nluModelUrls={{
-        // These files can made manually or
-        // exported from existing Alexa, Dialogflow, or
-        // Jovo models.
+      // These files can made manually or
```
Comment (space-pope) [inline README.md]:
```suggestion
      // These files can be made manually or
```
...though good luck with that :)

File: README.md [hunk 3]
```diff
@@ -467,91 +539,109 @@ wakeword and NLU models on device
...
 A key for a voice in Spokestack ASR, passed to Spokestack.synthesize
 
-### wakewordModelUrls
+### wakeword
+
+• `Optional` **wakeword**: SpokestackConfig[\"wakeword\"]
+
+_Defined in [src/SpokestackTray.tsx:275](https://github.com/spokestack/react-native-spokestack-tray/blob/fd8dea5/src/SpokestackTray.tsx#L275)_
+
+The NLU Tensorflow-Lite models (.tflite) for wakeword.
```
Comment (space-pope) [inline README.md]:
```suggestion
The NLU Tensorflow Lite models (.tflite) for wakeword.
```

File: README.md [hunk 4]
```diff
@@ -366,67 +417,84 @@ that gets played whenever the tray starts listening.
...
-### nluModelUrls
+### nlu
+
+• **nlu**: SpokestackConfig[\"nlu\"]
 
-• **nluModelUrls**: { metadata: string ; nlu: string ; vocab: string }
+_Defined in [src/SpokestackTray.tsx:128](https://github.com/spokestack/react-native-spokestack-tray/blob/fd8dea5/src/SpokestackTray.tsx#L128)_
 
-_Defined in [src/SpokestackTray.tsx:101](https://github.com/spokestack/react-native-spokestack-tray/blob/e00d4c9/src/SpokestackTray.tsx#L101)_
+The NLU Tensorflow-Lite model (.tflite), JSON metadata, and NLU vocabulary (.txt)
```
Comment (space-pope) [inline README.md]:
```suggestion
The NLU Tensorflow Lite model (.tflite), JSON metadata, and NLU vocabulary (.txt)
```

File: README.md [hunk 5]
```diff
@@ -467,91 +539,109 @@ wakeword and NLU models on device
...
+### wakeword
+
+• `Optional` **wakeword**: SpokestackConfig[\"wakeword\"]
+
+_Defined in [src/SpokestackTray.tsx:275](https://github.com/spokestack/react-native-spokestack-tray/blob/fd8dea5/src/SpokestackTray.tsx#L275)_
+
+The NLU Tensorflow-Lite models (.tflite) for wakeword.
 
-• `Optional` **wakewordModelUrls**: { detect: string ; encode: string ; filter: string }
+All 3 fields accepts 2 types of values.
```
Comment (space-pope) [inline README.md]:
```suggestion
All 3 fields accept 2 types of values.
```

File: README.md [hunk 6]
```diff
@@ -366,67 +417,84 @@ that gets played whenever the tray starts listening.
...
 
-_Defined in [src/SpokestackTray.tsx:101](https://github.com/spokestack/react-native-spokestack-tray/blob/e00d4c9/src/SpokestackTray.tsx#L101)_
+The NLU Tensorflow-Lite model (.tflite), JSON metadata, and NLU vocabulary (.txt)
 
-The URLs of your NLU model files.
-These models will be automatically downloaded
-the first time the app opens, and then saved.
-This is required for the NLU to work.
-See https://spokestack.io/docs/Concepts/nlu
+All 3 fields accepts 2 types of values.
```
Comment (space-pope) [inline README.md]:
```suggestion
All 3 fields accept 2 types of values.
```

File: README.md [hunk 7]
```diff
@@ -717,173 +761,112 @@ These functions are available as exports from react-native-spokestack-tray
...
 
 ```js
-import { isInitialized } from 'react-native-spokestack-tray'
-
-if (isInitialized()) {
-  // ...
-}
+useEffect(() => {
+  const listener = addEventListener('recognize', onRecognize)
+  // Unsubsribe by calling remove when components are unmounted
```
Comment (space-pope) [inline README.md]:
```suggestion
  // Unsubscribe by calling remove when components are unmounted
```

File: src/SpokestackTray.tsx [hunk 8]
```diff
@@ -92,17 +94,37 @@ interface Props {
     utterance?: string
   ) => IntentResult
   /**
-   * The URLs of your NLU model files.
-   * These models will be automatically downloaded
-   * the first time the app opens, and then saved.
-   * This is required for the NLU to work.
-   * See https://spokestack.io/docs/Concepts/nlu
+   * The NLU Tensorflow-Lite model (.tflite), JSON metadata, and NLU vocabulary (.txt)
```
Comment (space-pope) [inline src/SpokestackTray.tsx]:
```suggestion
   * The NLU Tensorflow Lite model (.tflite), JSON metadata, and NLU vocabulary (.txt)
```

File: src/SpokestackTray.tsx [hunk 9]
```diff
@@ -92,17 +94,37 @@ interface Props {
...
+   *
+   * All 3 fields accepts 2 types of values.
```
Comment (space-pope) [inline src/SpokestackTray.tsx]:
```suggestion
   * All 3 fields accept 2 types of values.
```

File: src/SpokestackTray.tsx [hunk 10]
```diff
@@ -213,18 +239,39 @@ interface Props {
   /** A key for a voice in Spokestack ASR, passed to Spokestack.synthesize */
   voice?: string
   /**
-   * The URLs of your wakeword model files.
-   * These models will be automatically downloaded
-   * the first time the app opens, and then saved.
-   * If no URLs are provided, the tray will default to
-   * the "Spokestack" wakeword.
-   * See https://spokestack.io/docs/Concepts/wakeword-models
+   * The NLU Tensorflow-Lite models (.tflite) for wakeword.
```
Comment (space-pope) [inline src/SpokestackTray.tsx]:
```suggestion
   * The NLU Tensorflow Lite models (.tflite) for wakeword.
```

File: src/SpokestackTray.tsx [hunk 11]
```diff
@@ -213,18 +239,39 @@ interface Props {
...
+   *
+   * All 3 fields accepts 2 types of values.
```
Comment (space-pope) [inline src/SpokestackTray.tsx]:
```suggestion
   * All 3 fields accept 2 types of values.
```

File: src/index.ts [hunk 12]
```diff
@@ -1,18 +1,68 @@
...
+export const isStarted = Spokestack.isStarted
+/**
+ * Bind to any event emitted by the native libraries
+ * The events are: "recognize", "partial_recognize", "error", "activate", "deactivate", and "timeout".
+ * See the bottom of the README.md for descriptions of the events.
+ *
+ * ```js
+ * useEffect(() => {
+ *   const listener = addEventListener('recognize', onRecognize)
+ *   // Unsubsribe by calling remove when components are unmounted
```
Comment (space-pope) [inline src/index.ts]:
```suggestion
 *   // Unsubscribe by calling remove when components are unmounted
```

File: src/SpokestackTray.tsx [hunk 13]
```diff
@@ -450,31 +475,41 @@ export default class SpokestackTray extends PureComponent<Props, State> {
...
     }
   }
 
-  private userSaid = ({ transcript }: SpokestackRecognizeEvent) => {
+  private onRecognize = async ({ transcript }: SpokestackRecognizeEvent) => {
+    const { editTranscript, onError } = this.props
+    console.log('[Spokestack onRecognize]:', transcript)
     this.utterance = transcript
     this.addBubble({ text: transcript, isLeft: false })
+    const edited = editTranscript(transcript)
```
Comment (space-pope) [inline src/SpokestackTray.tsx:485]: The Android version does the editing before the display in an attempt to make the conversation history look more like what the app wants. My view on it is that transcript editing is designed to correct ASR errors, and if the edited version is the source of truth for the NLU, it should also be that for the user looking back at their history to determine what went wrong in a misinterpreted request.

**Review by space-pope — COMMENTED** [review]
Mostly just trivial typos. I did have one question about transcript editing, though.

File: example/package.json [hunk 14]
```diff
@@ -14,16 +14,14 @@
...
     "babel-plugin-transform-inline-environment-variables": "^0.4.3",
     "react": "^17.0.1",
     "react-native": "0.63.4",
     "react-native-haptic-feedback": "^1.11.0",
     "react-native-linear-gradient": "^2.5.6",
     "react-native-permissions": "^3.0.0",
-    "react-native-spokestack": "4.0.0",
-    "react-native-video": "^5.1.0-alpha8",
-    "rn-fetch-blob": "^0.12.0"
+    "react-native-spokestack": "^5.0.0-beta.4",
```
Comment (noelweichbrodt) [inline example/package.json:23]: i assume this will be updated once rns hits 5.0.0?

**Review by noelweichbrodt — APPROVED** [review]

File: example/package.json (hunk 14 above)
Comment (timmywil) [reply example/package.json:23]: Yes, and the tray will be in beta until then.

File: src/SpokestackTray.tsx (hunk 13 above)
Comment (timmywil) [reply src/SpokestackTray.tsx:485]: Are you suggesting we display the edited transcript in the bubble?

File: src/SpokestackTray.tsx (hunk 8 above)
Comment (timmywil) [reply src/SpokestackTray.tsx]: You're funny. This is copy/paste from your edit in RNS.

Comment (space-pope) [reply src/SpokestackTray.tsx]: Wait, really? That doesn't sound like me; I've been on a crusade to make all our docs match their site - though that technically means I should've gone further here; it's officially `TendorFlow Lite`.

Now I have to go look at old prs to shame myself :).

File: src/SpokestackTray.tsx (hunk 13 above)
Comment (space-pope) [reply src/SpokestackTray.tsx:485]: I would, yeah, so the nlu and the user see the same thing.

Comment (timmywil) [reply src/SpokestackTray.tsx:485]: The only issue is when the format is set to something other than text, in which case the user would see ssml or markdown.

Comment (space-pope) [reply src/SpokestackTray.tsx:485]: Ssml or markdown would be for tts - this is pre-NLU, before any of that comes into play. Isn't the signature of `editTranscript` `string -> string`?

Comment (timmywil) [reply src/SpokestackTray.tsx:485]: I was thinking of synthesize. This is never ssml or markdown.

**Review by space-pope — APPROVED** [review]
LGTM. Good stuff.

**Pull request merged**